### PR TITLE
Blog - Feature deep dive: OpenSearch SQL Basic Queries

### LINKED ISSUE #612: Blog - OpenSearch SQL - Basic queries
**Describe the blog post your would like to write**
In this blog, we’ll explore some of the features and capabilities supported by OpenSearch SQL. We also briefly introduce how SQL engine works internally.

**What is the title of the blog post?**
OpenSearch SQL - Basic queries

**Who are the authors?**
penghuo, jadhanir

**What is the proposed posting date?**
No hard date, as soon as possible

### PR DESCRIPTION
**Pull request by penghuo**
Signed-off-by: Peng Huo <penghuo@gmail.com>

### Description
Blog - Feature deep dive: OpenSearch SQL Basic Queries
 
### Issues Resolved
#612 

### Check List
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

### CODE REVIEW

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+### Functions
+
+*“Find error logs where response code is 404 or 503, How many distinct host”* 
+You could run ODQL in Dashboards notebook also. There are 4 distinct hosts for about query,  result is in table format. More [aggregations](https://github.com/opensearch-project/sql/blob/main/docs/user/dql/aggregations.rst) and [functions](https://github.com/opensearch-project/sql/blob/main/docs/user/dql/functions.rst) are supported.
+![functions](/assets/media/blog-images/2022-01-30-opensearch-sql-basic-queries/functions.gif){: .img-fluid}
+### Explain
+
+*“How does ODQL execute in OpenSearch”*. Let’s explain an example query. OpenSearch SQL explain endpoint returns the query execution plan. For example, the above query two operators.
+
+* ProjectOperator, execute in coordinate node, read output from child operator and project originCountry and count fields.
```
Comment (stockholmux) [inline _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:42]: - "ProjectOperator, execute in coordinate node, read" ...  "originCountry and count" => "\`ProjectOperator\' , executes in the coordinating node reads" ... "\`originCountry\` and \`count\`"
- should it be "child operator" or "child operation"?

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+* FilterOperator, execute in coordinator node, read output from child operator and filter docs which count > 500.
```
Comment (stockholmux) [inline _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:43]: - "FilterOperator, execute in coordinator node, read output from child operator and filter docs which count > 500." => "\`FilterOperator\`, executes in coordinating node, reads output from child operator and filter docs which \`count > 500\`."

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+* OpenSearchIndexScan, it is a **DSL query** which will be executed in OpenSearch through search endpoint**.**
```
Comment (stockholmux) [inline _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:44]: - OpenSearchIndexScan it => \`OpenSearchIndexScan\'
- \*\*.\*\* => .

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+![explain](/assets/media/blog-images/2022-01-30-opensearch-sql-basic-queries/explain.gif){: .img-fluid}
+## Inside SQL Engines
+
+Internally, query will go through five major components in query engine. (1) **Language Processor** parse the query string by following the grammar and generate the AST (Abstract Syntax Tree). (2) **Core Engine** analyze, optimize the AST and build the Logical Plan. (3) **Storage Engine** is an pluggable component which provide the catalog schema and storage specified optimization and implementation. (4) **Execution Engine** schedule and execute the physical plan. (5) **Protocol** parse the request and control the response format.
```
Comment (stockholmux) [inline _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:49]: - Internally, query => Internally, a query
- in query engine => in the query engine
- parse the query => parses the query
- generate => generates
- "analyze, optimize the AST and build" => "analyzes and optimizes the AST and builds
- "is an pluggable component which provide" => "is a pluggable component which provides"
- "schedule and execute the physical plan" => "schedules and executes the physical plan"
- "parse the request and control" => "parses the request and controls"

**Review by stockholmux — CHANGES_REQUESTED** [review]
Content is pretty interesting - we need to tighten up the grammar and consistency a little bit.

I did a grammar pass and will probably need another one before publishing.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+categories: 
+  - feature
+
+---
+
+In this blog, we’ll explore some of the features and capabilities supported by OpenSearch SQL. We also briefly introduce how SQL engine works internally.
+
+## What is OpenSearch SQL?
+
+OpenSearch SQL is similar to the SELECT statement in ANSI SQL, but it is purely designed for analyzing data in OpenSearch. In this document, we refer to it as OSQL (short for OpenSearch Structured Query Language). In order to make it easier for traditional SQL customers to get started, we make the OSQL syntax conform to the subset of the ANSI SQL specification. In addition, we also disclosed the unique features of OpenSearch, for example, we support [full-text search](https://github.com/opensearch-project/sql/blob/main/docs/user/beyond/fulltext.rst).
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:17]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+## How to use OpenSearch SQL?
+
+For OpenSearch user, you could use the familiar REST interface `_plugin/sql` with your query in query body.
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:21]: Done. You are right, in the HTTP body.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+In this blog, we’ll explore some of the features and capabilities supported by OpenSearch SQL. We also briefly introduce how SQL engine works internally.
+
+## What is OpenSearch SQL?
+
+OpenSearch SQL is similar to the SELECT statement in ANSI SQL, but it is purely designed for analyzing data in OpenSearch. In this document, we refer to it as OSQL (short for OpenSearch Structured Query Language). In order to make it easier for traditional SQL customers to get started, we make the OSQL syntax conform to the subset of the ANSI SQL specification. In addition, we also disclosed the unique features of OpenSearch, for example, we support [full-text search](https://github.com/opensearch-project/sql/blob/main/docs/user/beyond/fulltext.rst).
+
+## How to use OpenSearch SQL?
+
+For OpenSearch user, you could use the familiar REST interface `_plugin/sql` with your query in query body. 
+For DBMS user, you could use [JDBC](https://github.com/opensearch-project/sql/tree/main/sql-jdbc)/[ODBC](https://github.com/opensearch-project/sql/tree/main/sql-odbc) driver to connect OpenSearch domain. You could download JDBC/ODBC driver. - add download link
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:22]: Agree. I use https://opensearch.org/artifacts#opensearch-sql-odbc for now. Will refresh it when download page is ready.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+For Dashboards user, you could use [OpenSearch Dashboard Query Workbench](https://opensearch.org/docs/latest/search-plugins/sql/workbench/) to easily run on-demand SQL queries and download results
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:23]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+### Basic Queries
+
+*“Find error logs where response code is 404 or 503”.* 
+You could run ODQL in Query Workbench. The result is in tabular format. You could download result in JSON, JDBC, CSV and Text format. Find more examples in [here](https://github.com/opensearch-project/sql/blob/main/docs/user/interfaces/protocol.rst).
+![basic queries](/assets/media/blog-images/2022-01-30-opensearch-sql-basic-queries/basic_queries.gif){: .img-fluid}
+### Functions
+
+*“Find error logs where response code is 404 or 503, How many distinct host”* 
+You could run ODQL in Dashboards notebook also. There are 4 distinct hosts for about query,  result is in table format. More [aggregations](https://github.com/opensearch-project/sql/blob/main/docs/user/dql/aggregations.rst) and [functions](https://github.com/opensearch-project/sql/blob/main/docs/user/dql/functions.rst) are supported.
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:36]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+![functions](/assets/media/blog-images/2022-01-30-opensearch-sql-basic-queries/functions.gif){: .img-fluid}
+### Explain
+
+*“How does ODQL execute in OpenSearch”*. Let’s explain an example query. OpenSearch SQL explain endpoint returns the query execution plan. For example, the above query two operators.
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:40]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown (hunk 1 above)
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:42]: Done.
child operator is more accurate.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown (hunk 2 above)
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:43]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown (hunk 3 above)
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:44]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown (hunk 4 above)
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:49]: Done. Thanks a lot.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+authors: 
+  - penghuo
+  - jadhanir
+date: 2022-01-23
+categories: 
+  - feature
+
+---
+
+In this blog, we’ll explore some of the features and capabilities supported by OpenSearch SQL. We also briefly introduce how SQL engine works internally.
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown:13]: Done.

File: _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+For DBMS user, you could use [JDBC](https://github.com/opensearch-project/sql/tree/main/sql-jdbc)/[ODBC](https://github.com/opensearch-project/sql/tree/main/sql-odbc) driver to connect OpenSearch domain. You could download JDBC/ODBC driver. - add download link
+For Dashboards user, you could use [OpenSearch Dashboard Query Workbench](https://opensearch.org/docs/latest/search-plugins/sql/workbench/) to easily run on-demand SQL queries and download results
+An additional OpenSearch SQL CLI tool is provided for interactive SQL execution.
+
+## Some Examples
+
+### Basic Queries
+
+*“Find error logs where response code is 404 or 503”.* 
+You could run ODQL in Query Workbench. The result is in tabular format. You could download result in JSON, JDBC, CSV and Text format. Find more examples in [here](https://github.com/opensearch-project/sql/blob/main/docs/user/interfaces/protocol.rst).
```
[the remark this answers is not in the document]
Comment (penghuo) [reply _posts/2022-01-23-feature-deep-dive-opensearch-sql-basic-queries.markdown]: Done.

**Comment by penghuo** [comment]
@stockholmux Thanks for the feedback, Very helpful. I resolve comments which I already addressed, but I am leaving comment 2 unresolved. I'll update the JDBC/ODBC link as it becomes available.

**Comment by stockholmux** [comment]
Sounds good @penghuo. I'll wait on this until we get the ODBC/JDBC drivers up.

**Comment by stockholmux** [comment]
@penghuo FYI - ODBC/JDBC drivers are on the downloads page. I'd be glad to push this live tomorrow if we have that updated :)

**Comment by penghuo** [comment]
> @penghuo FYI - ODBC/JDBC drivers are on the downloads page. I'd be glad to push this live tomorrow if we have that updated :)

Thanks!!!. I updated the ODBC/JDBC drivers link.

Should I change the release-date? Currently, I use `2022-01-23`.

**Review by stockholmux — DISMISSED** [review]
LGTM

**Review by stockholmux — APPROVED** [review]
LGTM

**Pull request merged**